src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Receivers.kt

```diff
@@ -150,11 +152,87 @@ class MarkdownTaskReceiver(
+    // TODO: fixedThreadPool? if so, how many threads?
+    private val executor = Executors.newFixedThreadPool(4)
+    private val currentTasks = ConcurrentHashMap<String, Future<Boolean>>()
 
     override fun processMessage(message: PubsubMessage): Boolean {
-        val request = CompileRequest.parseFrom(message.data)
+        val request = Request.parseFrom(message.data)
+
+        when (request.typeCase) {
+            Request.TypeCase.COMPILE -> return processCompile(request.compile)
+            Request.TypeCase.CANCEL  -> return processCancel(request.cancel)
+            else -> {
+                LOGGER.error("Request type is not set")
+            }
+        }
+
+        return true
+    }
+
+    private fun processCancel(request: CancelRequestProto): Boolean {
+        LOGGER.debug("Canceling the task with id = {}", request.taskId)
+        var status = CompilerFleet.Cancel.Status.FAILED // init value
+        val elapsedTime = measureTimeMillis {
+            status = cancelTask(request)
+        }
+
+        val data = CompilerFleet.Cancel.newBuilder()
+                .setTaskId(request.taskId)
+                .setStatus(status)
+                .setCpuTime(elapsedTime.toInt())
+                .build()
+                .toByteString()
+        val onPublishFailureCallback = {
+            LOGGER.info("Publish cancel task ${request.taskId} response failed with code ${StatusCode.FAILURE}")
+        }
+
+        resultPublisher.publish(data, onPublishFailureCallback)
+        return true
+    }
+
+    private fun cancelTask(request: CancelRequestProto): CompilerFleet.Cancel.Status {
+        return if (!currentTasks.contains(request.taskId)) {
+            CompilerFleet.Cancel.Status.NOT_FOUND
+        } else {
+            val result = currentTasks[request.taskId]?.cancel(true)!!
+            currentTasks.remove(request.taskId)
+            // will canceling the thread solve the problem?
+            // another idea: two tasks maps of each processing stage
+            // 1) tasks are converting to latex at the moment
+            // 2) tasks are pushed to the texbe
+            //
+            // 1) we stop by stopping the cmd process
+            // 2) we stop by sending a cancel request to the texbe
+
+            if (result) {
+                CompilerFleet.Cancel.Status.OK
+            } else {
+                CompilerFleet.Cancel.Status.FAILED
+            }
+        }
+    }
+
+    private fun processCompile(request: CompileRequest): Boolean {
         LOGGER.debug("Converting Markdown to tex: {}", request.mainFileName)
 
+        if (currentTasks.contains(request.id)) {
+            LOGGER.debug("Task with id {} is already executing", request.id)
+            return true
+        }
+
+        val future: Future<Boolean> = executor.submit(Callable {
+            return@Callable processTask(request)
+        })
+
+        currentTasks[request.id] = future
+        val result = future.get()
```

Comment:
This essentially makes our task blocking one and all those executors useless (you wait anyway until task completes => you do not accept other tasks, including cancel requests)

Reply:
Yes, this is weird :)
Fixed